### PR TITLE
Prepare for non-Base58 addresses

### PR DESCRIPTION
**Pull request by sipa**
This patch removes the need for the intermediary Base58 type `CBitcoinAddress`, by providing {`Encode`,`Decode`,`IsValid`}`Destination` functions that directly operate on the conversion between `std::string`s and `CTxDestination`.

As a side, it also fixes a number of indentation issues, and removes probably several unnecessary implicit `CTxDestination`<->`CBitcoinAddress` conversions.

This change is far from complete. In follow-ups I'd like to:
* Split off the specific address and key encoding logic from base58.h, and move it to a address.h or so.
* Replace `CTxDestination` with a non-`boost::variant` version (which can be more efficient as `boost::variant` allocates everything on the heap, and remove the need for `boost::get<...>` and `IsValidDestination` calls everywhere).
* Do the same for `CBitcoinSecret`, `CBitcoinExtKey`, and `CBitcoinExtPubKey`.

However, I've tried to keep this patch to be minimally invasive, but still enough to support non-Base58 addresses. Perhaps a smaller patch is possible to hack Bech32 support into `CBitcoinAddress`, but I would consider that a move in the wrong direction.

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -275,24 +275,25 @@ UniValue setaccount(const JSONRPCRequest& request)
...
         strAccount = AccountFromValue(request.params[1]);
 
     // Only add the account if the address is yours.
-    if (IsMine(*pwallet, address.Get())) {
+    if (IsMine(*pwallet, dest)) {
         // Detect when changing the account of an address that is the 'unused current key' of another account:
-        if (pwallet->mapAddressBook.count(address.Get())) {
-            std::string strOldAccount = pwallet->mapAddressBook[address.Get()].name;
-            if (address == GetAccountAddress(pwallet, strOldAccount)) {
+        if (pwallet->mapAddressBook.count(dest)) {
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Nit, kind of unrelated, use `.find()` to avoid 2nd lookup below.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -451,9 +454,10 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
 
     LOCK2(cs_main, pwallet->cs_wallet);
 
-    CBitcoinAddress address(request.params[0].get_str());
-    if (!address.IsValid())
+    CTxDestination dest = DecodeDestination(request.params[0].get_str());
+    if (!IsValidDestination(dest)) {
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid Bitcoin address");
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Follow up, should the error message be `Invalid destination address`?

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -2695,18 +2704,20 @@ UniValue listunspent(const JSONRPCRequest& request)
...
-            if (!address.IsValid())
-                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, std::string("Invalid Bitcoin address: ")+input.get_str());
-            if (setAddress.count(address))
-                throw JSONRPCError(RPC_INVALID_PARAMETER, std::string("Invalid parameter, duplicated address: ")+input.get_str());
-           setAddress.insert(address);
+            CTxDestination dest = DecodeDestination(input.get_str());
+            if (!IsValidDestination(dest)) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, std::string("Invalid Bitcoin address: ") + input.get_str());
+            }
+            if (destinations.count(dest)) {
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Replace with `.insert().second` like above.

**Review by promag — COMMENTED** [review]
There is this pattern
```cpp
DecodeDestination();
if (!IsValidDestination()) {
```
What about:
```cpp
CTxDestination dest;
if (!DecodeDestination(str, dest)) {
```

Also, to avoid more usages of `CBitcoinAddress`, add commit to move it to `base58.cpp`?

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (sipa) [inline src/wallet/rpcwallet.cpp]: An address is just the string-encoded form of a CTxDestination, so no. I think the "destination" name should be internal.

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (sipa) [inline src/wallet/rpcwallet.cpp]: Trying to keep this as much of a straightforward change as possible, as it may need backporting.

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (promag) [inline src/wallet/rpcwallet.cpp]: Off topic, just `Invalid address` as it seems irrelevant to have `Bitcoin` in the error message?

Comment (sipa) [inline src/wallet/rpcwallet.cpp]: Oh, sure!

**Comment by sipa** [comment]
> What about:
> 
> ```c++
> CTxDestination dest;
> if (!DecodeDestination(str, dest)) {
> ```

There are plenty of cases where the result of decoding can just be passed to something else, and `Function(DecodeDestination(addr))` is far cleaner than `CTxDestination dest; DecodeDestination(addr, dest); Function(dest);`, so I prefer to keep that possible.

The current pattern is indeed a bit verbose, but once we rewrite `CTxDestination` to be a native type (rather than a `boost::variant`), it could at least simplify to `dest.IsValid()` or so.

**Comment by promag** [comment]
@sipa sure, I just wanted to question that "style" because it's used elsewhere.

ACK 8363a2c.

**Comment by jonasschnelli** [comment]
utACK 8363a2ccaffba793068db4f97184121730707104

**Comment by laanwj** [comment]
Added 0.15.1 milestone

**Review by kallewoof — APPROVED** [review]
utACK 8363a2ccaffba793068db4f97184121730707104
LGTM. I skipped over QT stuff as I don't know it very well.

File: src/script/standard.cpp [hunk 4]
```diff
@@ -317,3 +317,7 @@ CScript GetScriptForWitness(const CScript& redeemscript)
     ret << OP_0 << ToByteVector(hash);
     return ret;
 }
+
+bool IsValidDestination(const CTxDestination& dest) {
+    return dest.which() != 0;
```
Comment (theuni) [inline src/script/standard.cpp]: Maybe add a quick comment that this is CNoDestination so it comes up when grepping for future changes.

File: src/script/standard.h [hunk 5]
```diff
@@ -77,10 +77,13 @@ class CNoDestination {
  *  * CNoDestination: no destination set
  *  * CKeyID: TX_PUBKEYHASH destination
  *  * CScriptID: TX_SCRIPTHASH destination
- *  A CTxDestination is the internal data type encoded in a CBitcoinAddress
+ *  A CTxDestination is the internal data type encoded in a bitcoin address
  */
 typedef boost::variant<CNoDestination, CKeyID, CScriptID> CTxDestination;
 
+/** Check whether a CTxDestination is a CNoDestination. */
```
Comment (theuni) [inline src/script/standard.h]: got it. ignore last nit.

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -529,16 +533,16 @@ UniValue listaddressgroupings(const JSONRPCRequest& request)
...
         UniValue jsonGrouping(UniValue::VARR);
-        for (CTxDestination address : grouping)
+        for (const CTxDestination& address : grouping)
         {
             UniValue addressInfo(UniValue::VARR);
-            addressInfo.push_back(CBitcoinAddress(address).ToString());
+            addressInfo.push_back(EncodeDestination(address));
             addressInfo.push_back(ValueFromAmount(balances[address]));
             {
-                if (pwallet->mapAddressBook.find(CBitcoinAddress(address).Get()) != pwallet->mapAddressBook.end()) {
```
Comment (theuni) [inline src/wallet/rpcwallet.cpp]: heh.

**Review by theuni — COMMENTED** [review]
utACK other than the keyID issue.

File: src/qt/signverifymessagedialog.cpp [hunk 7]
```diff
@@ -117,16 +117,14 @@ void SignVerifyMessageDialog::on_signMessageButton_SM_clicked()
...
+    if (!IsValidDestination(destination)) {
         ui->statusLabel_SM->setStyleSheet("QLabel { color: red; }");
         ui->statusLabel_SM->setText(tr("The entered address is invalid.") + QString(" ") + tr("Please check the address and try again."));
         return;
     }
-    CKeyID keyID;
-    if (!addr.GetKeyID(keyID))
-    {
+    const CKeyID* keyID = boost::get<CKeyID>(&destination);
+    if (keyID) {
```
Comment (promag) [inline src/qt/signverifymessagedialog.cpp]: Good catch.

File: src/qt/addresstablemodel.cpp [hunk 8]
```diff
@@ -82,14 +82,14 @@ class AddressTablePriv
             LOCK(wallet->cs_wallet);
             for (const std::pair<CTxDestination, CAddressBookData>& item : wallet->mapAddressBook)
             {
-                const CBitcoinAddress& address = item.first;
```
Comment (sipa) [inline src/qt/addresstablemodel.cpp]: It was creating a temporary of type CBitcoinAddress by converting the CTxDestination initializer, using the implicit `CBitcoinAddress(const CTxDestination &dest)` constructor. That temporary's lifetime was then extended by taking a reference to it, named `address`.

File: src/wallet/rpcwallet.cpp (hunk 6 above)
Comment (sipa) [inline src/wallet/rpcwallet.cpp]: It took me a while to figure out what this `CBitcoinAddress` was doing. Answer: nothing at all.

File: src/qt/signverifymessagedialog.cpp (hunk 7 above)
Comment (sipa) [inline src/qt/signverifymessagedialog.cpp]: Wow, thanks :S

File: src/qt/addresstablemodel.cpp (hunk 8 above)
Comment (theuni) [inline src/qt/addresstablemodel.cpp]: Interesting, I didn't know an implicit conversion could be done when extending the lifetime of a const ref. Scary!

**Comment by sipa** [comment]
Removed `CBitcoinAddress` entirely, using a commit from #11167.

**Review by promag — COMMENTED** [review]
Since this is the base for Bech32, maybe move everything related to `CTxDestination` out of base58, including tests, to a new place. Also move `CTxDestination` definition and related functions from `src/script/standard.h`?

At least should be done in a follow up?

File: src/base58.cpp [hunk 9]
```diff
@@ -12,6 +12,7 @@
 #include <string.h>
 #include <vector>
 #include <string>
+#include <algorithm>
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/base58.cpp]: Fixed.

File: src/qt/sendcoinsdialog.cpp [hunk 10]
```diff
@@ -777,21 +777,20 @@ void SendCoinsDialog::coinControlChangeEdited(const QString& text)
...
 
-        CBitcoinAddress addr = CBitcoinAddress(text.toStdString());
+        CTxDestination destination = DecodeDestination(text.toStdString());
 
         if (text.isEmpty()) // Nothing entered
         {
             ui->labelCoinControlChangeLabel->setText("");
         }
-        else if (!addr.IsValid()) // Invalid address
+        else if (!IsValidDestination(destination)) // Invalid address
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/qt/sendcoinsdialog.cpp]: Not going to reflow this entire function in a PR intended for backporting.

File: src/rpc/misc.cpp [hunk 11]
```diff
@@ -207,15 +208,14 @@ UniValue validateaddress(const JSONRPCRequest& request)
...
+    CTxDestination dest = DecodeDestination(request.params[0].get_str());
+    bool isValid = IsValidDestination(dest);
 
     UniValue ret(UniValue::VOBJ);
     ret.push_back(Pair("isvalid", isValid));
     if (isValid)
     {
-        CTxDestination dest = address.Get();
-        std::string currentAddress = address.ToString();
+        std::string currentAddress = EncodeDestination(dest);
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/rpc/misc.cpp]: There may be reasons not to do that. Bech32 addresses (added in a later PR) can be uppercase or lowercase, but we always want to store and look up in lowercase form. This is an easy way of doing that, and we should probably check that it's done consistently elsewhere too.

**Comment by sipa** [comment]
> Since this is the base for Bech32, maybe move everything related to CTxDestination out of base58, including tests, to a new place.

Yes, read the PR description. I plan to do this, but not in a patch that's intended for backporting.

> Also move `CTxDestination` definition and related functions from `src/script/standard.h`?

Why? To where?

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/11117/commits/87d5c28583374f1be6273c3620b29b0a505e0191

though I do not know the QT stuff well.

File: src/rpc/misc.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (promag) [reply src/rpc/misc.cpp]: Agree, I'm +1 for clear data validation and/or sanitization. Although `encode(decode(data))` is a obscure way of doing it, not to mention `encode(decode(data))` should equal `data`?

**Comment by promag** [comment]
> Removed CBitcoinAddress entirely, using a commit from #11167.

@sipa missing this commit as `CBitcoinAddress` is still around.

**Comment by sipa** [comment]
@promag Yes, I removed it again.

This PR is just the part that is likely to break easily and need non-trivial rebasing.

File: src/qt/transactiondesc.cpp [hunk 12]
```diff
@@ -91,9 +91,8 @@ QString TransactionDesc::toHTML(CWallet *wallet, CWalletTx &wtx, TransactionReco
         if (nNet > 0)
         {
             // Credit
-            if (CBitcoinAddress(rec->address).IsValid())
-            {
-                CTxDestination address = CBitcoinAddress(rec->address).Get();
+            if (IsValidDestination(rec->address)) {
+                CTxDestination address = DecodeDestination(rec->address);
```
Comment (promag) [inline src/qt/transactiondesc.cpp]: This incur in a double decoding since above `IsValidDestination` also decodes. Maybe swap:
```cpp
CTxDestination address = DecodeDestination(rec->address);
if (IsValidDestination(address)) {
```

File: src/script/standard.cpp (hunk 4 above)
Comment (promag) [reply src/script/standard.cpp]: Could use `!(operator==)` instead?

**Comment by laanwj** [comment]
Not sure if we support this version, but I get the following compile error with clang 3.5.0:
```
/home/user/src/bitcoin/src/qt/signverifymessagedialog.cpp:235:40: error: invalid operands to binary expression ('CTxDestination' (aka 'variant<CNoDestination, CKeyID, CScriptID>') and 'CTxDestination')
    if (CTxDestination(pubkey.GetID()) != destination) {
```
Then a lot of note: candidate function not viable like:
```
/usr/include/x86_64-linux-gnu/qt5/QtCore/qchar.h:544:13: note: candidate function not viable: no known conversion from 'CTxDestination' (aka 'variant<CNoDestination, CKeyID, CScriptID>') to 'QChar' for 1st argument
inline bool operator!=(QChar c1, QChar c2) { return c1.unicode() != c2.unicode(); }
```

**Comment by JeremyRubin** [comment]
@laanwj 

changing it to 


    if (!(CTxDestination(pubkey.GetID()) == destination)) {

probably fixes it. Ran into this while writing https://github.com/bitcoin/bitcoin/pull/9991

**Comment by laanwj** [comment]
I confirm that @JeremyRubin's workaround makes it compile.

**Comment by sipa** [comment]
@JeremyRubin Fixed

**Comment by morcos** [comment]
utACK eebedca

**Comment by sipa** [comment]
@laanwj Changed to `IsValidDestinationString`.

**Comment by laanwj** [comment]
utACK 864cd2787

File: src/wallet/rpcwallet.cpp [hunk 13]
```diff
@@ -454,9 +457,10 @@ UniValue sendtoaddress(const JSONRPCRequest& request)
     ObserveSafeMode();
     LOCK2(cs_main, pwallet->cs_wallet);
 
-    CBitcoinAddress address(request.params[0].get_str());
-    if (!address.IsValid())
-        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid Bitcoin address");
+    CTxDestination dest = DecodeDestination(request.params[0].get_str());
+    if (!IsValidDestination(dest)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid address");
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: nit: Any reason why this was changed from "Invalid Bitcoin address" to "Invalid address" and the other places left as is. This is part of a error string, so shouldn't break anything I guess.

**Review by maflcko — COMMENTED** [review]
(really late) post-merge utACK 864cd278741ab0e17e7d93eedee39417d4797b37

File: src/wallet/rpcwallet.cpp (hunk 13 above)
Comment (promag) [reply src/wallet/rpcwallet.cpp]: @achow101 regarding the need for error testing.

**Comment by MarcoFalke** [comment]
Removing from backport
